wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -98,56 +98,75 @@ private[bitcoins] trait TransactionProcessing extends WalletLogger {
-                blockHashOpt = Some(block.blockHeader.hash.flip),
-                newTags = Vector.empty,
-                receivedSpendingInfoDbsOpt = Some(receivedSpendingInfoDbs),
-                spentSpendingInfoDbsOpt = Some(spentSpendingInfo)
-              )
-            }
-            _ = {
-              //need to look if a received utxo is spent in the same block
-              //if so, we need to update our cachedSpentF
-              val spentInSameBlock: Vector[SpendingInfoDb] = {
-                processTxResult.updatedIncoming.filter { spendingInfoDb =>
-                  block.transactions.exists(
-                    _.inputs.exists(
-                      _.previousOutput == spendingInfoDb.outPoint))
-                }
+    val blockHashOpt = Some(block.blockHeader.hash.flip)
+    val resultF: Future[Future[Wallet]] = for {
+      //map on these first so we don't have to call
+      //.map everytime we iterate through a tx
+      //which is costly (thread overhead)
+      receivedSpendingInfoDbsOpt <- cachedReceivedOptF
+      spentSpendingInfoDbs <- spentSpendingInfoDbsF
+    } yield {
+      //we need to keep a cache of spentSpendingInfoDb
+      //for the case where we receive & then spend that
+      //same utxo in the same block
+      var cachedSpentOpt: Option[Vector[SpendingInfoDb]] = {
+        Some(spentSpendingInfoDbs)
+      }
+      val blockInputs = block.transactions.flatMap(_.inputs)
+      val wallet: Future[Wallet] = {
+        block.transactions.foldLeft(Future.successful(this)) {
+          (walletF, transaction) =>
+            for {
+              wallet <- walletF
+              processTxResult <- {
+                wallet.processTransactionImpl(
+                  transaction = transaction,
+                  blockHashOpt = blockHashOpt,
+                  newTags = Vector.empty,
+                  receivedSpendingInfoDbsOpt = receivedSpendingInfoDbsOpt,
+                  spentSpendingInfoDbsOpt = cachedSpentOpt
+                )
               }
+              _ = {
+                //need to look if a received utxo is spent in the same block
+                //if so, we need to update our cachedSpentF
+                val spentInSameBlock: Vector[SpendingInfoDb] = {
+                  processTxResult.updatedIncoming.filter { spendingInfoDb =>
+                    blockInputs.exists(
+                      _.previousOutput == spendingInfoDb.outPoint)
+                  }
+                }
 
-              //add it to the cache
-              cachedSpentF =
-                Future.successful(spentSpendingInfo ++ spentInSameBlock)
+                //add it to the cache
+                val newCachedSpentOpt = {
+                  cachedSpentOpt match {
+                    case Some(spentSpendingInfo) =>
+                      Some(spentSpendingInfo ++ spentInSameBlock)
+                    case None =>
+                      Some(spentInSameBlock)
+                  }
+                }
```

Comment:
would this be more effienct?

```scala
val newSpent = newCachedSpentOpt.getOrElse(Vector.empty)
val newCachedSpentOpt = cachedSpentOpt.map(_ ++ newSpent)
```

Reply:
I'm not sure to be honest. I think it comes down to what is more efficiently implemented by the Scala language

1. the `match` keyword along with deconstructing `case` statements
2. `Option.map()`and `.getOrElse()`. 

I'm guessing it is a minuscule performance difference either way.